web/src/components/vis/ForceDirectedGraph.vue

```diff
@@ -119,13 +119,26 @@ export default {
       const nodes = select(this.$refs.svg).select('g.nodes').selectAll('g').select('circle');
       const edges = select(this.$refs.svg).select('g.edges').selectAll('g').select('line');
       // hide nodes that have been deleted from search results
-      nodes.style('visibility', (node) => (excluded.has(node.id) ? 'hidden' : ''));
+      nodes.style('visibility', (node) => (excluded.has(node.id) ? 'hidden' : 'visible'));
       edges.style('visibility', (edge) => (excluded.has(edge.source.id) || excluded.has(edge.target.id) ? 'hidden' : 'visible'));
       this.showNodesWithinPathLength(this.search, this.visibleNodes);
     },
     visibleNodes(visibleNodes) {
       this.showNodesWithinPathLength(this.search, visibleNodes);
     },
+    nodes(newNodes) {
+      // hide/unhide nodes based on node filter
+      const newNodeSet = new Set(newNodes.map((node) => node.id)); // convert to set for fast lookup
+      const nodes = select(this.$refs.svg).select('g.nodes').selectAll('g').select('circle');
+      const edges = select(this.$refs.svg).select('g.edges').selectAll('g').select('line');
+      nodes.style('visibility', (node) => (newNodeSet.has(node.id) ? 'visible' : 'hidden'));
+      edges.style('visibility', (edge) => (newNodeSet.has(edge.source.id) && newNodeSet.has(edge.target.id) ? 'visible' : 'hidden'));
+      nodes.style('color', (node) => node.color);
+    },
+  },
+  updated() {
+    const nodes = select(this.$refs.svg).select('g.nodes').selectAll('g').select('circle');
+    nodes.style('fill', (node, index) => (this.search.includes(node.id) ? 'red' : this.nodes[index].color));
```

Comment:
I'm losing track of what's going on here.  Why is updated needed instead of using a watcher?  Using updated obfuscates the reader's ability to figure out what is intended to cause this side-effect.

Reply:
The nodes need to be recolored after each rerender of the graph to preserve the selected color nodes. You're right, a watcher would be much better, just switched it.